monomerdb/localdb/localdb.go

```diff
@@ -0,0 +1,464 @@
+func (db *DB) Rollback(unsafe, safe, finalized common.Hash) error {
+	return db.updateIndexed(func(b *pebble.Batch) (err error) {
+		unsafeHeightBytesValue, closer, err := get(b, bucketHeightByHash.Key(unsafe.Bytes()))
+		if err != nil {
+			return fmt.Errorf("get height by hash %s: %w", unsafe, err)
+		}
+		defer func() {
+			err = wrapCloseErr(err, closer)
+		}()
+		unsafeHeight := endian.Uint64(unsafeHeightBytesValue)
+		firstHeightBytesToDelete := marshalUint64(unsafeHeight + 1)
+
+		// Set the new height.
+		if err := b.Set(heightKey, unsafeHeightBytesValue, nil); err != nil {
+			return fmt.Errorf("set height: %v", err)
+		}
+
+		// Delete headers.
+		firstHeaderToDelete := bucketHeaderByHeight.Key(firstHeightBytesToDelete)
+		nextBucket := (bucketHeaderByHeight + 1).Key()
+		iter, err := b.NewIter(&pebble.IterOptions{
+			LowerBound: firstHeaderToDelete,
+			UpperBound: nextBucket,
+		})
+		if err != nil {
+			return fmt.Errorf("new bucketHeaderByHeight iterator: %v", err)
+		}
+		defer func() {
+			err = wrapCloseErr(err, iter)
+		}()
+		header := new(monomer.Header)
+		for iter.First(); iter.Valid(); iter.Next() {
+			value, err := iter.ValueAndErr()
+			if err != nil {
+				return fmt.Errorf("get header value from iterator: %v", err)
+			}
+			if err := cbor.Unmarshal(value, &header); err != nil {
+				return fmt.Errorf("unmarshal header from cbor: %v", err)
+			}
+			if err := b.Delete(bucketHeightByHash.Key(header.Hash.Bytes()), nil); err != nil {
+				return fmt.Errorf("delete height by hash %s: %v", header.Hash, err)
+			}
+		}
+		if err := b.DeleteRange(firstHeaderToDelete, nextBucket, nil); err != nil {
+			return fmt.Errorf("delete range of headers: %v", err)
+		}
+
+		// Delete transactions.
+		firstTxToDelete := bucketTxByHeightAndIndex.Key(firstHeightBytesToDelete)
+		nextBucket = (bucketTxByHeightAndIndex + 1).Key()
+		iter, err = b.NewIter(&pebble.IterOptions{
+			LowerBound: firstTxToDelete,
+			UpperBound: nextBucket,
+		})
+		if err != nil {
+			return fmt.Errorf("new bucketTxByHeightAndIndex iterator: %v", err)
+		}
+		defer func() {
+			err = wrapCloseErr(err, iter)
+		}()
+		for iter.First(); iter.Valid(); iter.Next() {
+			value, err := iter.ValueAndErr()
+			if err != nil {
+				return fmt.Errorf("get tx value from iterator: %v", err)
+			}
+			hash := bfttypes.Tx(value).Hash()
+			if err := b.Delete(bucketTxHeightAndIndexByHash.Key(hash), nil); err != nil {
+				return fmt.Errorf("delete tx height and index by hash %v: %v", hash, err)
+			}
+		}
+		if err := b.DeleteRange(
+			bucketTxHeightAndIndexByHash.Key(firstHeightBytesToDelete),
+			marshalUint64(uint64(bucketTxHeightAndIndexByHash)+1),
+			nil,
+		); err != nil {
+			return fmt.Errorf("delete range of txs: %v", err)
+		}
+
+		if err := updateLabels(b, unsafe, safe, finalized); err != nil {
+			return fmt.Errorf("update labels: %v", err)
+		}
+		return nil
+	})
+}
+
+func (db *DB) HeadHeader() (*monomer.Header, error) {
```

Comment:
This actually was an important change: we were leaking an iterator (not closing) in `localdb.DB.Rollback` because we were reassigning it before closing it. Closing the db caught the issue!

Reply:
woops, meant to leave the above response on a separate comment.

I ended up deciding against renaming `Head`->`Latest` for now, since `latest` can be a synonym for `unsafe` in Ethereum-speak (especially in the JSON-RPC API). To avoid confusion between the head block in the db and the `unsafe` block, I'll leave it as it is for now.